### PR TITLE
Add message tree state model

### PR DESCRIPTION
**Pull request by danielpatrickhug**
This commit is part of issue #37 
1. added the new SQL model MessageTreeState
2. Added reference to backend/__init__
3. ran alembic revision script
4. ran pre-commit

This PR won't break anything if merged. However, message_tree_state isn't functional i.e. adding a new message won't initialize a tree state.

### CODE REVIEW

File: backend/oasst_backend/models/message_tree_state.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from typing import Optional
+from uuid import UUID, uuid4
+
+import sqlalchemy as sa
+import sqlalchemy.dialects.postgresql as pg
+from sqlalchemy import false
+from sqlmodel import Field, Index, SQLModel
+
+INITIAL = "initial"
+BREEDING_PHASE = "breeding_phase"
```
Comment (andreaskoepf) [inline backend/oasst_backend/models/message_tree_state.py:11]: these should probably be put into a str-enum (see other cases where enums are used in the prj)

File: backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "message_tree_state",
+        sa.Column("id", postgresql.UUID(as_uuid=True), server_default=sa.text("gen_random_uuid()"), nullable=False),
+        sa.Column("created_date", sa.DateTime(), server_default=sa.text("CURRENT_TIMESTAMP"), nullable=False),
+        sa.Column("deleted", sa.Boolean(), server_default=sa.text("false"), nullable=False),
```
Comment (andreaskoepf) [inline backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py:26]: not sure about the deleted flag here, that should already be part of message, right?

File: backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "message_tree_state",
+        sa.Column("id", postgresql.UUID(as_uuid=True), server_default=sa.text("gen_random_uuid()"), nullable=False),
+        sa.Column("created_date", sa.DateTime(), server_default=sa.text("CURRENT_TIMESTAMP"), nullable=False),
```
Comment (andreaskoepf) [inline backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py:25]: created_date is also already part of message .. so would be redundant. I think we should remove/not add  it here.

**Review by andreaskoepf — CHANGES_REQUESTED** [review]
please let us not add created_date and deleted to the message_tree_state table.

File: backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py (hunk 2 above)
Comment (danielpatrickhug) [reply backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py:26]: I apologize, I made the proper changes to remove it. this pull request now adds two alembic revisions. should I restart the pull request to resolve this?

File: backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py (hunk 3 above)
Comment (danielpatrickhug) [reply backend/alembic/versions/2023_01_05_1144-d4161e384f83_added_messagetreestate_table.py:25]: removed

File: backend/oasst_backend/models/message_tree_state.py (hunk 1 above)
Comment (danielpatrickhug) [reply backend/oasst_backend/models/message_tree_state.py:9]: okay, Ill make the changes

Comment (danielpatrickhug) [reply backend/oasst_backend/models/message_tree_state.py]: ok rewrote the Enum class for states and updated VALID_STATES

File: backend/oasst_backend/models/message_tree_state.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from uuid import UUID, uuid4
+
+import sqlalchemy as sa
+import sqlalchemy.dialects.postgresql as pg
+from sqlmodel import Field, Index, SQLModel
+
+# The types of States a message tree can have.
+
+
+class States(Enum):
```
Comment (yk) [inline backend/oasst_backend/models/message_tree_state.py:12]:
```suggestion
class States(str, Enum):
```

File: backend/oasst_backend/models/message_tree_state.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+
+VALID_STATES = (
+    States.INITIAL,
+    States.BREEDING_PHASE,
+    States.RANKING_PHASE,
+    States.READY_FOR_SCORING,
+    States.CHILDREN_SCORED,
+    States.FINAL,
+)
```
Comment (yk) [inline backend/oasst_backend/models/message_tree_state.py:28]: could this be replaced by either `tuple(States)` or `tuple(s.value for s in States)` ?

**Review by yk — COMMENTED** [review]
Thank you very much, just left some minor comments

File: backend/oasst_backend/models/message_tree_state.py (hunk 4 above)
Comment (danielpatrickhug) [reply backend/oasst_backend/models/message_tree_state.py:12]: No problem

File: backend/oasst_backend/models/message_tree_state.py (hunk 5 above)
Comment (danielpatrickhug) [reply backend/oasst_backend/models/message_tree_state.py:28]: Sure

**Pull request merged**